src/cockpit/channels/metrics.py

```diff
@@ -15,23 +15,179 @@
 import logging
+from typing import Any, Dict, List, Set, NamedTuple, Optional
+from collections import defaultdict
 
-from ..channel import Channel
+from ..channel import AsyncChannel
+from ..samples import SAMPLERS, BaseSampler, SampleDescription
 
 logger = logging.getLogger(__name__)
 
 
-class MetricsChannel(Channel):
+class MetricInfo(NamedTuple):
+    name: str
+    derive: Optional[str]
+    units: str
+    semantics: str
+    instanced: bool
+
+
+class InternalMetricsChannel(AsyncChannel):
     payload = 'metrics1'
+    restrictions = {'source': 'internal'}
+
+    metrics: List[MetricInfo]
+    samplers: Set
+
+    samples: Dict[str, Any]
+    old_samples: Dict[str, Any]
+
+    samplers_cache: Optional[Dict[str, BaseSampler]] = None
+    description_cache: Optional[Dict[str, SampleDescription]] = None
+
+    interval: int = 1000
+    need_meta: bool = False
+    last_timestamp: int = 0
+    next_timestamp: int = 0
+
+    @classmethod
+    def ensure_samplers(cls):
+        if cls.samplers_cache is None:
+            cls.samplers_cache = {}
+            cls.descriptions_cache = {}
+            for sampler in SAMPLERS:
+                print(sampler)
+                for name, desc in sampler.descriptions.items():
+                    cls.samplers_cache[name] = sampler
+                    cls.descriptions_cache[name] = desc
+
+    def parse_options(self, options):
+        sampler_classes = set()
+        logger.debug("metrics internal open: %s, channel: %s", options, self.channel)
+
+        interval = options.get('interval', self.interval)
+        if interval <= 0 or interval > sys.maxsize:
+            logger.error('invalid "interval" value: %s', interval)
+            self.close(problem="protocol-error", message='invalid interval option')
+            return False
+
+        self.interval = interval
+
+        metrics = options.get('metrics')
+        if not metrics or not isinstance(metrics, list):
+            logger.error('invalid "metrics" value: %s', metrics)
+            self.close(problem='metrics is not an list')
+            return False
+
+        for metric in metrics:
+            # validate it's an object
+            name = metric.get('name')
+            units = metric.get('units')
+            derive = metric.get('derive')
+
+            desc = self.descriptions_cache.get(name)
+            if not desc:
+                logger.error('unsupported metric: %s', name)
+                self.close(problem='not-supported',
+                           message='invalid metric')
+                return False
+
+            if units and units != desc.units:
+                self.close(problem='protocol-error',
+                           message=f'{name} has units {desc.units}, not {units}')
+                return False
+
+            sampler_classes.add(self.samplers_cache.get(name))
+            self.metrics.append(MetricInfo(name=name, derive=derive, semantics=desc.semantics, units=desc.units, instanced=desc.instanced))
```

Comment:
I wonder if it wouldn't make sense to have a `MetricInfo` constructor which took a `SampleDescription` and a `derive` parameter (since the `derive` parameter seems to be the only thing here which isn't part of the description).

You could also make good use of some syntax like

```python3
MetricInfo(**desc._asdict(), derive=derive)
```

See below for comments about how to deal with the `name` field.

One more note: `MetricInfo` seems to be more or less the same as `SampleDescription` with some extra fields.  Does a subtype relationship make sense here?  Or maybe a containment relationship?  I played a bit with namedtuple subtyping, and I'd lean towards containment based on my findings...

I don't like the duplication of the 3 or 4 fields, basically...